internal_worker.go

```diff
@@ -405,53 +396,81 @@ func (th *hostEnvImpl) AddActivityRegistrationInterceptor(i interceptorFn) {
-	th.activityFuncMap = make(map[string]interface{}) // clear map
+	th.activityFuncMap = make(map[string]activity) // clear map
 	th.Unlock()
-	for w, f := range funcMapCopy {
-		intw, intf := i(w, f)
-		th.Lock()
-		th.activityFuncMap[intw] = intf
-		th.Unlock()
+	for w, a := range funcMapCopy {
+		intw, intf := i(w, a.GetFunction())
+		th.addActivity(intw, &activityExecutor{intw, intf})
 	}
 }
 
 func (th *hostEnvImpl) RegisterWorkflow(af interface{}) error {
+	return th.RegisterWorkflowWithOptions(af, RegisterWorkflowOptions{})
+}
+
+func (th *hostEnvImpl) RegisterWorkflowWithOptions(
+	af interface{},
+	options RegisterWorkflowOptions,
+) error {
 	// Validate that it is a function
 	fnType := reflect.TypeOf(af)
 	if err := validateFnFormat(fnType, true); err != nil {
 		return err
 	}
-	// Check if already registered
 	fnName := getFunctionName(af)
-	if _, ok := th.getWorkflowFn(fnName); ok {
-		return fmt.Errorf("workflow type \"%v\" is already registered", fnName)
+	alias := options.Name
+	registerName := fnName
+	if len(alias) > 0 {
+		registerName = alias
+	}
+	// Check if already registered
+	if _, ok := th.getWorkflowFn(registerName); ok {
+		return fmt.Errorf("workflow name \"%v\" is already registered", registerName)
 	}
 	// Register args with encoding.
 	if err := th.registerEncodingTypes(fnType); err != nil {
 		return err
 	}
-	fnName, af = th.invokeInterceptors(fnName, af, th.workflowRegistrationInterceptors)
-	th.addWorkflowFn(fnName, af)
+	registerName, af = th.invokeInterceptors(registerName, af, th.workflowRegistrationInterceptors)
+	th.addWorkflowFn(registerName, af)
+	if len(alias) > 0 {
+		th.addWorkflowAlias(fnName, alias)
+	}
 	return nil
 }
 
 func (th *hostEnvImpl) RegisterActivity(af interface{}) error {
+	return th.RegisterActivityWithOptions(af, RegisterActivityOptions{})
+}
+
+func (th *hostEnvImpl) RegisterActivityWithOptions(
+	af interface{},
+	options RegisterActivityOptions,
+) error {
 	// Validate that it is a function
 	fnType := reflect.TypeOf(af)
 	if err := validateFnFormat(fnType, false); err != nil {
 		return err
 	}
-	// Check if already registered
 	fnName := getFunctionName(af)
-	if _, ok := th.getActivityFn(fnName); ok {
-		return fmt.Errorf("activity type \"%v\" is already registered", fnName)
+	alias := options.Name
+	registerName := fnName
+	if len(alias) > 0 {
+		registerName = alias
+	}
+	// Check if already registered
+	if _, ok := th.getActivityFn(registerName); ok {
+		return fmt.Errorf("activity type \"%v\" is already registered", registerName)
 	}
 	// Register args with encoding.
 	if err := th.registerEncodingTypes(fnType); err != nil {
 		return err
 	}
-	fnName, af = th.invokeInterceptors(fnName, af, th.activityRegistrationInterceptors)
-	th.addActivityFn(fnName, af)
+	registerName, af = th.invokeInterceptors(registerName, af, th.activityRegistrationInterceptors)
+	th.addActivityFn(registerName, af)
```

Comment:
I think we might have problem with this interceptor. We transform the rergisterName when register, but we do not transform when invoke. cc @mfateev  (Note: this is not related to this PR)